### PR TITLE
Detect and configure the replicated SDK chart

### PR DESCRIPTION
**Pull request by cbodonnell**
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/upstream/write.go [hunk 1]
```diff
@@ -198,3 +232,557 @@ func maybeEncryptIdentityConfig(identityConfig *kotsv1beta1.IdentityConfig) ([]b
...
+
+type ReplicatedValuesOptions struct {
+	LicenseYAML []byte
+	ClusterID   string
+	AppID       string
+}
+
+func addReplicatedValuesForSDK(doc *yaml.Node, opts ReplicatedValuesOptions, replicatedSDKSubchartName string, isReplicatedSDK bool) error {
+	replicatedValues := map[string]interface{}{
+		"license":               string(opts.LicenseYAML),
```
Comment (cbodonnell) [inline pkg/upstream/write.go:530]: should only need to do this in airgap

File: pkg/upstream/write.go [hunk 2]
```diff
@@ -198,3 +232,557 @@ func maybeEncryptIdentityConfig(identityConfig *kotsv1beta1.IdentityConfig) ([]b
...
+		for i, n := range targetNode.Content {
+			if n.Value == "replicated" {
+				targetNode = targetNode.Content[i+1]
+				hasGlobalReplicated = true
+				break
+			}
+		}
+	}
+
+	replicatedValues := map[string]interface{}{
```
Comment (cbodonnell) [inline pkg/upstream/write.go:609]: should only need to do this in airgap

File: pkg/upstream/write.go [hunk 3]
```diff
@@ -100,7 +115,25 @@ func WriteUpstream(u *types.Upstream, options types.WriteOptions) error {
...
-		if err := ioutil.WriteFile(fileRenderPath, file.Content, 0644); err != nil {
+		if archives.IsTGZ(file.Content) {
+			// this is a helm chart, so we need to check if it is or contains the replicated-sdk
+			reader := bytes.NewReader(file.Content)
+			replicatedSDKChartName, isReplicatedSDK, err := FindReplicatedSDKChart(reader, u.ReplicatedChartNames)
+			if err != nil {
+				return errors.Wrap(err, "failed to find replicated-sdk subchart")
+			}
+
+			if replicatedSDKChartName != "" {
```
Comment (sgalsaleh) [inline pkg/upstream/write.go:126]: why only if it's a subchart?

File: pkg/upstream/write.go [hunk 4]
```diff
@@ -198,3 +232,557 @@ func maybeEncryptIdentityConfig(identityConfig *kotsv1beta1.IdentityConfig) ([]b
 
 	return b.Bytes(), nil
 }
+
+// FindReplicatedSDKChart will look for a chart with the name "replicated" or "replicated-sdk" in the archive
```
Comment (sgalsaleh) [inline pkg/upstream/write.go:236]: should update this comment to reflect that it uses `replicatedChartNames` and not hardcoded values
```suggestion
// FindReplicatedSDKChart will look for a chart with the name "replicated" or "replicated-sdk" in the archive
```

File: pkg/upstream/write.go [hunk 5]
```diff
@@ -198,3 +232,557 @@ func maybeEncryptIdentityConfig(identityConfig *kotsv1beta1.IdentityConfig) ([]b
...
+					ChartName string `json:"name" yaml:"name"`
+				}{}
+				if err := yaml.Unmarshal(content, &chartInfo); err != nil {
+					return "", false, errors.Wrapf(err, "failed to unmarshal %s", header.Name)
+				}
+
+				if replicatedChartNamesMap[chartInfo.ChartName] {
+					// check if the sdk is the parent chart or a subchart based on the path
+					replicatedSDKChartName := chartInfo.ChartName
+					isReplicatedSDK := !strings.Contains(filepath.Dir(header.Name), string(os.PathSeparator))
```
Comment (sgalsaleh) [inline pkg/upstream/write.go:278]: wouldn't there be a top-level directory before the filename? e.g. `replicated/Chart.yaml`.

File: pkg/upstream/write.go [hunk 6]
```diff
@@ -198,3 +232,557 @@ func maybeEncryptIdentityConfig(identityConfig *kotsv1beta1.IdentityConfig) ([]b
...
+			content, err := io.ReadAll(tarReader)
+			if err != nil {
+				finalErr = errors.Wrap(err, "failed to read file")
+				return
+			}
+
+			if filepath.Base(header.Name) == "values.yaml" {
+				// only get the values.yaml from the top level chart
+				p := filepath.Dir(header.Name)
+				if !strings.Contains(p, string(os.PathSeparator)) {
```
Comment (sgalsaleh) [inline pkg/upstream/write.go:371]: ditto

File: pkg/upstream/write.go [hunk 7]
```diff
@@ -198,3 +232,557 @@ func maybeEncryptIdentityConfig(identityConfig *kotsv1beta1.IdentityConfig) ([]b
...
+	if hasGlobalReplicated || hasGlobal {
+		targetNode.Content = kotsutil.MergeYAMLNodes(targetNode.Content, additionalNode.Content[0].Content)
+	} else {
+		targetNode.Content = append(targetNode.Content, additionalNode.Content[0].Content...)
+	}
+
+	return nil
+}
+
+func convertLicenseFieldsToYamlNodes(licenseFields map[string]LicenseField) ([]*yaml.Node, error) {
```
Comment (sgalsaleh) [inline pkg/upstream/write.go:650]: can't you do a `yaml.Marshal` then `yaml.Unmarshal` into nodes?

File: pkg/upstream/write.go (hunk 3 above)
Comment (cbodonnell) [reply pkg/upstream/write.go:126]: this should have a value if it's found at all, regardless of whether it's a parent or subchart. the `isReplicatedSDK` variable (now renamed to `isSubchart`) is what makes the distinction

File: pkg/upstream/write.go (hunk 7 above)
Comment (sgalsaleh) [reply pkg/upstream/write.go:653]: agreed.

**Review by sgalsaleh — APPROVED** [review]

**Pull request merged**